x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportCloseJobAction.java

```diff
@@ -249,21 +250,97 @@ void stopDatafeedsIfNecessary(OpenAndClosingIds jobIds, boolean isForce, TimeVal
                     if (runningDatafeedIds.isEmpty()) {
                         listener.onResponse(false);
                     } else {
-                        StopDatafeedAction.Request request = new StopDatafeedAction.Request(String.join(",", runningDatafeedIds));
-                        request.setForce(isForce);
-                        request.setStopTimeout(timeout);
-                        ClientHelper.executeAsyncWithOrigin(
-                            client,
-                            ClientHelper.ML_ORIGIN,
-                            StopDatafeedAction.INSTANCE,
-                            request,
-                            ActionListener.wrap(r -> listener.onResponse(r.isStopped()), listener::onFailure));
+                        if (isForce) {
+                            // A datafeed with an end time will gracefully close its job when it stops even if it was force stopped.
+                            // If we didn't do anything about this then it would turn requests to force close jobs into normal close
+                            // requests for those datafeeds, which is undesirable - the caller specifically asked for the job to be
+                            // closed forcefully, skipping the final state persistence to save time.  Therefore, before stopping the
+                            // datafeeds in this case we isolate them.  An isolated datafeed will NOT close its associated job under
+                            // any circumstances.  The downside of doing this is that if the stop datafeeds call fails then this API
+                            // will not have closed any jobs, but will have isolated one or more datafeeds, so the failure will have
+                            // an unexpected side effect.  Hopefully the project to combine jobs and datafeeds will be able to improve
+                            // on this.
+                            isolateDatafeeds(jobIds.openJobIds, runningDatafeedIds, ActionListener.wrap(
+                                r -> stopDatafeeds(runningDatafeedIds, true, timeout, listener),
+                                listener::onFailure
```

Comment:
This is never called, looking at the isolateDatafeeds function, it never calls listener.onFailure. Might be good to make that clear.